src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,146 @@
+            elements = newElements;
+            size++;
+        } else {
+            growIfNeed();
+            T[] newElements = (T[]) new Object[elements.length];
+            System.arraycopy(elements, 0, newElements, 0, index);
+            newElements[index] = value;
+            System.arraycopy(elements, index, newElements, index + 1, size - index + 1);
+            elements = newElements;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            growIfNeed();
+            elements[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
+
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
 
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
+
+        T oldValue = elements[index];
+        if (index == size - 1) {
+            elements[index] = null;
+        } else if (index == 0) {
+            T[] newElements = (T[]) new Object[elements.length];
+            System.arraycopy(elements, 1, newElements, 0, size);
+            elements = newElements;
+        } else {
+            T[] newElements = (T[]) new Object[elements.length];
+            System.arraycopy(elements, 0, newElements, 0, index);
+            System.arraycopy(elements, index + 1, newElements, index, size - index);
+            elements = newElements;
+        }
+
+        size--;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = findElement(element);
+        if (index == -1) {
+            throw new NoSuchElementException("Element: " + element + " not found");
+        }
+
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private int findElement(T element) {
+        for (int i = 0; i < size; i++) {
+            if (element == elements[i] || (elements[i] != null && elements[i].equals(element))) {
+                return i;
+            }
+        }
+
+        return NOT_FOUND;
+    }
+
+    private void growIfNeed() {
+        int valuesLength = elements.length;
+        if (size == valuesLength) {
+            T[] extendedValues = (T[]) new Object[valuesLength + (int) (valuesLength * 1.5)];
```

Comment:
extract magic value to constant
```suggestion
            T[] extendedValues = (T[]) new Object[valuesLength + (int) (valuesLength * GROW_FACTOR)];
```